File: spec/features/course/registration_spec.rb
# frozen_string_literal: true
require 'rails_helper'

RSpec.feature 'Courses: Registration' do
  let!(:instance) { Instance.default }

  with_tenant(:instance) do
    let(:course) { create(:course) }
    let(:user) { create(:user) }
    before { login_as(user, scope: :user) }

    context 'when the course has unconfirmed invitations' do
      let!(:invitation) { create(:course_user_invitation, course: course) }

      scenario 'Users can register course using invitation code', js: true do
        visit course_path(course)

        # No code input
        find('#register-button').click
        expect(page).
          to have_selector('div.Toastify__toast-body', text: 'Please enter an invitation code')

        # Enter wrong registration code
        fill_in 'registration-code', with: 'defintielyTheWrongCode'
        find('#register-button').click
        expect(page).
          to have_selector('div.Toastify__toast-body', text: 'Your code is incorrect')

        # Correct code
        fill_in 'registration-code', with: invitation.invitation_key
        find('#register-button').click
      end
    end

    context 'when the course allows enrol requests' do
      let(:course) { create(:course, :enrollable) }

      scenario 'Users can create and cancel enrol requests', js: true do
        visit course_path(course)

        expect(page).to have_text(course.description)

        expect(ActionMailer::Base.deliveries.count).to eq(0)
        find('#submit-enrol-request-button').click
        expect(page).to have_selector('div.Toastify__toast-body', text: 'Your enrol request has been submitted.')
        expect(ActionMailer::Base.deliveries.count).not_to eq(0)

        # Cancel request
        find('#cancel-enrol-request-button').click
        expect(page).to have_selector('div.Toastify__toast-body', text: 'Your enrol request has been cancelled.')
      end

      context 'when the user has been enrolled' do
        let!(:enrolled_student) { create(:course_student, course: course, user: user) }

        scenario 'user cannot de-register or re-register for the course', js: true do
          visit course_path(course)
          expect(page).not_to have_selector('#submit-enrol-request-button')
          expect(page).not_to have_selector('#cancel-enrol-request-button')
        end
      end
    end
  end
end

File: tests/tests/courses/registration.spec.ts
import {
  test,
  expect,
  manufacture,
  getLastSentEmail,
  clearEmails,
} from 'helpers';

test.describe('with an unconfirmed invitation', () => {
  let course: { id: number };
  let invitation;

  test.beforeEach(async () => {
    course = await manufacture({ course: { traits: ['published']} });
    invitation = await manufacture({
      course_user_invitation: { course_id: course.id },
    });
  });

  test('can register with invitation code', async ({ authedPage: page }) => {
    await page.goto(`/courses/${course.id}`);

    await expect(page.getByText(invitation.name)).not.toBeVisible();

    const registerButton = page.getByRole('button', { name: 'Register' });
    await registerButton.click();

    await expect(page.getByText('enter an invitation code')).toBeVisible();

    const invitationCodeField = page.getByLabel('Invitation code');
    await invitationCodeField.fill('definitelyTheWrongCode');
    await registerButton.click();

    await expect(page.getByText('code is incorrect')).toBeVisible();

    await invitationCodeField.fill(invitation.invitation_key);
    await registerButton.click();

    await expect(page.getByText(invitation.name)).toBeVisible();
  });
});

test.describe('allows enrol requests', () => {
  let course: { id: number };

  test.beforeEach(async () => {
    course = await manufacture({ course: { traits: ['published', 'enrollable'] } });

    await clearEmails();
  });

  test('can request', async ({ authedPage: page }) => {
    await page.goto(`/courses/${course.id}`);

    await expect(page.getByText('Description')).toBeVisible();
    await expect(page.getByText('Instructors', { exact: true })).toBeVisible();

    page.getByRole('button', { name: 'Request to enrol' }).click();

    await expect(
      page.getByText('enrol request has been submitted'),
    ).toBeVisible();

    const notificationEmail = await getLastSentEmail();
    expect(notificationEmail?.subject).toContain('Enrol Request');

    page.getByRole('button', { name: 'Cancel request' }).click();

    await expect(
      page.getByText('enrol request has been cancelled'),
    ).toBeVisible();
  });

  test('cannot request if already enrolled', async ({ authedPage: page }) => {
    const student = await manufacture({
      course_student: { course_id: course.id, user_id: page.user.id },
    });

    await page.goto(`/courses/${course.id}`);

    await expect(page.getByText(student.name)).toBeVisible();

    await expect(
      page.getByRole('button', { name: 'Request to enrol' }),
    ).not.toBeVisible();

    await expect(
      page.getByRole('button', { name: 'Cancel request' }),
    ).not.toBeVisible();
  });
});
